Rollforward of #1476

ChadKillingsworth opened the pull request:
Rollforward #1476 which was rolled back as part of 641cf61

Partially addresses #1625

This PR may have been the biggest reason for the rollback. In particular, it may be subject to the same restrictions as #1573 which caused a circular dependency problem with internal Google builds. If so, I can quite easily refactor this in the same manner to avoid that issue.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="35" align="absmiddle" alt="Review on Reviewable"/>](https://reviewable.io/reviews/google/closure-compiler/1628)
<!-- Reviewable:end -->

ChadKillingsworth commented:
Rebased again to fix the merge conflict.

ChadKillingsworth commented:
Rebased again to fix the merge conflict. Keeping up with a rapidly moving target is difficult.

blickly commented:
One issue I noticed with this is that it causes us to call JsFileParser (in particular the setIncludeGoogBase method) more often, which is not yet implemented in the GWT version of the compiler.

ChadKillingsworth commented:
@Dominator008 Do you know why this isn't implemented for GWT?

https://github.com/google/closure-compiler/blob/master/src/com/google/javascript/jscomp/gwt/super/com/google/javascript/jscomp/deps/JsFileParser.java#L27

That's the line that's failing.

Dominator008 commented:
@ChadKillingsworth I think that particular method can be implemented.

ChadKillingsworth commented:
@blickly @Dominator008 Requesting the provided/required symbol lists was re-parsing the files. But module re-writing had already notified the compiler input class of those symbols. I added some new methods to CompilerInput to prevent a full regeneration of dependency information when all we really want was the information it already knew about.

@blickly If this seems to work, I'll squash the commits and rebase to the top of the tree.

Dominator008 commented:
@ChadKillingsworth https://github.com/google/closure-compiler/blob/master/src/com/google/javascript/jscomp/gwt/super/com/google/javascript/jscomp/deps/JsFileParser.java#L30 Will this `parseFile()` be called? This will be more work to implement since it uses regexes.

ChadKillingsworth commented:
@Dominator008 No - I completely sidestepped the entire `JsFileParser` usage with this change. When `input.getProvides()` is called, it creates a `JsFileParser` instance to scan the code to find all `goog.require` and `goog.provide` statements. That is completely un-necessary since module rewriting already reports those as they are generated. See https://github.com/google/closure-compiler/blob/master/src/com/google/javascript/jscomp/ProcessCommonJSModules.java#L358-L359

Dominator008 commented:
@ChadKillingsworth I see, so we don't have to provide a GWT version of `setIncludeGoogBase()` for now.

ChadKillingsworth commented:
@Dominator008 Correct - I found this out after my question earlier today. I had to trace through the code to discover how this all worked.

blickly commented:
Found another issue here with creating URLs with invalid characters:

Here's an example stack:
```
java.lang.IllegalArgumentException: Illegal character in scheme name at index 9: synthetic_module_$48:mod:Stub
		at java.net.URI.create(URI.java:852)
		at com.google.javascript.jscomp.ModuleIdentifier.forFile(ModuleIdentifier.java:97)
		at com.google.javascript.jscomp.Compiler.parseInputs(Compiler.java:1369)
		at com.google.javascript.jscomp.Compiler.parse(Compiler.java:730)
		at com.google.javascript.jscomp.JSCompiler.parse(JSCompiler.java:156)
		at com.google.javascript.jscomp.Compiler.compileInternal(Compiler.java:691)
		at com.google.javascript.jscomp.Compiler.access$000(Compiler.java:85)
		at com.google.javascript.jscomp.Compiler$2.call(Compiler.java:662)
		at com.google.javascript.jscomp.Compiler$2.call(Compiler.java:659)
		at com.google.javascript.jscomp.CompilerExecutor$2.call(CompilerExecutor.java:93)
		at java.util.concurrent.FutureTask.run(FutureTask.java:266)
		at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
		at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
		at java.lang.Thread.run(Thread.java:745)
```

MatrixFrog commented:
I hit a similar issue when merging another PR. Look at the comments on cl/116700722

ChadKillingsworth commented:
[ES6ModuleLoader.java](https://github.com/google/closure-compiler/pull/1628/files#diff-51277464fa4940394ed596618d6d0b25R97) already had an implementation for invalid characters. If this still isn't enough, I can simply add a try/catch block around this call and skip any files whose name won't convert to a URI.

I rebased this to the top of the tree to handle the merge conflicts. I also discovered that the tests were treating this like an optimization/checks pass. I switched the tests to mirror the ES6 module tests so that the rewriting occurs first thing.

blickly commented on src/com/google/javascript/jscomp/ProcessEs6Modules.java:
```diff
@@ -401,11 +402,11 @@ private void visitScript(NodeTraversal t, Node script) {
   }
 
-  private static void checkStrictModeDirective(NodeTraversal t, Node n) {
+  private static void setStrictModeDirective(NodeTraversal t, Node n) {
```
Looks like the NodeTraversal argument is no longer needed here

ChadKillingsworth replied:
Fixed

MatrixFrog commented:
This is a bit weird because ES6 transpilation will now behave differently on the same file when you run it as part of a full compilation, vs. in a file-by-file way.

We need to make sure we know what the implications of that are. Maybe we can just say that if your workflow requires file-by-file transpilation, and you need to import a module that exports nothing, then you have to add `export default undefined;` or something like that. But it would be nice not to have such a requirement.

ChadKillingsworth commented:
@MatrixFrog Currently we already mention that you do have to export something for a module to be treated as such. There is no way to determine if a file is a module unless you know that:

 1. It exports something or
 2. something else imports it

The file-by-file transpilation concept is incompatible with modules who are imported for side effects only. I know of no way around that. See #1220

MatrixFrog commented:
Another option we've considered a few times is some kind of explicit annotation in the `@fileoverview` block to tell the compiler "treat this file as a module" but that's nearly as hacky as `export default undefined`. So I guess I'm okay with this as long as it's documented.

Dominator008 commented:
@MatrixFrog @ChadKillingsworth Will something like an `'isEs6Module` directive at the beginning of the file work? That's also hacky but clearer I guess.

ChadKillingsworth commented:
@Dominator008 At that point there is no difference from exporting undefined. You lose the ability to treat the file as both script or module.
